### ISSUE TITLE
How can I run tests in parallel?

### ISSUE THREAD

**Issue by alexbjorlig** [microsoft/playwright#18875]
I'm trying to migrate our integration API tests from jest to Playwright. In that process, I would like to ask a question about running tests in parallel.

Let's say I have 100 test files. I would like 10 files to be executed in parallel.
Each running test should have it's own isolated server running.

My questions:

- Where do I start the server - in the playwright.webServer config?
- Do I simply start the webServer in the beforeAll hook?

Thank you so much.

Extra info

- tests are written in typescript
- our API is written in Sveltekit and can therefore start locally with the vite dev server
- We use the MongoDB in-memory server as backend when doing tests
- When I'm done with the migration of integration tests, I would like to migrate our Cypress e2e tests also.

**Comment by pavelfeldman** [comment]
There are several strategies for implementing this. If you'd like 10 parallel tracks, it means that you'll be running 10 parallel workers. Note that typically parallelism is determined based on the resources rather than based on the number of tests. I.e. 1 worker on CI, 6 workers on MacBook, 25 workers on Xeon workstation. You can always add `--workers=4` in the end of the Playwright command line to override it.

When you are saying that "each test should have its own isolated server running", do you mean that you start and stop server for each test (server per test) or that one server is used by one test at a time (server per worker). It sounds from the context that you are talking about the latter. If you are interested in the server-per-worker, the most elegant solution is to implement it as a worker fixture. Worker fixture will run once for each worker and then will tear down when the worker is done. Read this page https://playwright.dev/docs/next/test-fixtures and then refer to https://playwright.dev/docs/next/test-fixtures#worker-scoped-fixtures for the worker fixture.

**Comment by pavelfeldman** [comment]
https://playwright.dev/docs/next/test-advanced#workerinfo-object

**Comment by alexbjorlig** [comment]
This is amazing - thanks for providing the links. Yes - I'm 95% certain that I'm looking for server per worker.

As long as I can specify that tests inside a worker should **not** run in parallel. I'm reading the links, thanks for providing the info 🙏

**Comment by pavelfeldman** [comment]
Closing as per above, please feel free to open a new issue if this does not cover your use case.

**Comment by alexbjorlig** [comment]
Ok, I now have some issues with the "server per worker" strategy. I migrated all my tests to playwright, but to my big surprise, I got more than 5 workers. I start the process with `--workers 5`, so I would only expect the **workerIndex** to be in range 0-4? But in the console I can see that the worker fixture is called many times.

I this because workers are not re-used between test files - or what I'm I missing?

**worker-setup.ts**
```typescript
export const test = base.extend<{}, { server: string }>({
    server: [
        async ({}, use, workerInfo) => {
            const { workerIndex } = workerInfo;
            // Start mongodb-memory-server
            // Start vite (sveltekit) dev server (and save serverUri)...

            console.log(`workerIndex=${workerIndex}`);
            await use(serverUri);
            // clean...
        },
        { scope: 'worker', auto: true },
    ],
});
```

**typical test**
```
import { test } from '../worker-setup';

test.describe('pagination-recs', () => {

   test('something should work 1 ', // ...)
   test('something should work 2 ', // ...)
});
```

**playwright.config.ts**
```typescript
const config: PlaywrightTestConfig = {
    testDir: 'dist',
    /* Maximum time one test can run for. */
    timeout: 30 * 1000,
    expect: {
        /**
         * Maximum time expect() should wait for the condition to be met.
         * For example in `await expect(locator).toHaveText();`
         */
        timeout: 5000,
    },
    /* Run tests in files in parallel */
    fullyParallel: false,
    /* Fail the build on CI if you accidentally left test.only in the source code. */
    forbidOnly: !!process.env.CI,
    /* Retry on CI only */
    retries: process.env.CI ? 2 : 0,
    /* Opt out of parallel tests on CI. */
    workers: process.env.CI ? 1 : undefined,
    /* Reporter to use. See https://playwright.dev/docs/test-reporters */
    reporter: 'html',
    /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
    use: {
        /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
        actionTimeout: 0,
        /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
        trace: 'on-first-retry',
    },
};
```


**start command**
```
playwright test --workers 5
```

**Comment by alexbjorlig** [comment]
Update - the fault here was actually workers failing and automatically restarting to try again.
